File: corners.m
close all;
%{
Robot= [-50 -100 0 1;150 0 0 1;-50 100 0 1]';% The Robot icon is a triangle
patch(Robot(1,:), Robot(2,:), 'b')
%}
hold on;

% Inicialització de punts
%{
points = []
for i=0:10
    points = [points;i,10] 
end

for i=0:10
    points = [points;10,i] 
end

scatter(points(:,1),points(:,2))
%}
%figure;

%%{
fid = fopen("vista_robot.txt");
fitxer = fscanf(fid, '%s');
linies = split(fitxer, ';');
x = [];
y = [];
for i=1:length(linies)-1
    xy = split(linies(i), ',');
    x = [x;str2double(xy(1,1))];
    y = [y;str2double(xy(2,1))];
end
scatter(x,y,".");
fclose(fid);
axis([-2000 2000 -2000 2000])
points = [x,y];
%%}
% Detecció de corners
cantonada = [];
xy = [];
w = 30; %30;
max_dist = 5000; %100000;
for i=1:1:length(points)-w
    dist = sqrt((points(i+w,1)-points(i,1))^2 + (points(i+w,2)-points(i,2))^2);
    if dist < max_dist
        ux = points(i+w,1) - points(i+w/2,1);
        uy = points(i+w,2) - points(i+w/2,2);
        vx = points(i+w/2,1) - points(i,1);
        vy = points(i+w/2,2) - points(i,2);
        pas2 = (ux*vx) + (uy*vy);
        pas3 = sqrt(ux^2 + uy^2) * sqrt(vx^2 + vy^2);
        pas1 = pas2/pas3;
        beta = acos(pas1)*(180/pi);
        if (beta > 80 && beta < 100) || (beta > 260 && beta < 280)
            xy = [xy; points(i+3,:)];
        end
    end
end

% xy

%%{
for i=1:1:length(xy)
    cont = 0;
    for j=1:1:length(xy)
        dist = sqrt((xy(i,1)-xy(j,1))^2 + (xy(i,2)-xy(j,2))^2);
        if dist < 500 && i ~= j && dist ~= 0
            cont = cont + 1;
        end
        %xy(i,:)
        %xy(j,:)
    end
    xy(i,:), cont
    if cont < 4
        xy(i,:) = nan;
    end
end
%%}

scatter(0,0,"g");
if length(xy) >= 1
    scatter(xy(:,1),xy(:,2),'r')
end

hold off;

File: corners.py
import math


def clean(punts):
    llista = []
    for i in range(290, 359):  # 290, 359
        if punts[i][0] != 0 and punts[i][1] != 0:
            llista.append(punts[i])
    for i in range(0, 90):  # 0 , 70
        if punts[i][0] != 0 and punts[i][1] != 0:
            llista.append(punts[i])
    print(llista)
    return llista


def corners(points):
    xy = []
    w = 16  # 6
    max_dist = 10000000  # 5000;
    for i in range(0, len(points)-w):
        dist = math.sqrt((points[i+w][0] - points[i][0]) ** 2 + (points[i+w][1] - points[i][1]) ** 2)
        if dist < max_dist:
            ux = points[i+w][0] - points[i+w/2][0]
            uy = points[i+w][1] - points[i + w / 2][1]
            vx = points[i + w / 2][0] - points[i][0]
            vy = points[i + w / 2][1] - points[i][1]
            pas2 = (ux * vx) + (uy * vy)
            pas3 = math.sqrt(ux ** 2 + uy ** 2) * math.sqrt(vx ** 2 + vy ** 2)
            pas1 = pas2 / pas3
            beta = math.acos(pas1) * (180 / math.pi)
            if (beta > 60 and beta < 120) or (beta > 240 and beta < 300):
                xy.append(points[i+w/2])
    return xy


def cluster(points):
    for i in points:
        cont = 0
        for j in points:
            dist = math.sqrt((i[0] - j[0]) ** 2 + (i[1] - j[1]) ** 2)
            if dist < 10000 and i[0] != j[0] and i[1] != j[1] and dist != 0:
                cont = cont + 1
        if cont < 3:  # 4
            points.remove(i)
    return points



